### ISSUE TITLE
API key ?

### ISSUE THREAD

**Issue by jfayad** [openpim/server#45]
I couldn't find anyway to get an API key that would be usable for API consumption.

Right now, it looks like we need to login a user to get a token that we can use for the API, even if that works, the issue is with the token validity duration that needs a new login every once in a while...

**Comment by SebastiaanDenHertog** [comment]
Hi @jfayad, 

You can use the graphql endpoint signin for this:
signIn(login: String!, password: String!): LoginResponse!

the request looks like:
query: "mutation {signIn(login: \"yourusername\", password: \"yourpassword\") { token, user {id internalId login name email roles tenantId options props external} }}"

and you get if its correct:
{
    "data": {
        "signIn": {
            "token": "yourToken",
            "user": {
                "id": "",
                "internalId": ,
                "login": "",
                "name": "",
                "email": "",
                "roles": [
                    ""
                ],
                "tenantId": "default",
                "options": [],
                "props": {},
                "external": false
            }
        }
    }
}

then you can use your token for API and graphql calls in you put it in the header as {X-Token:"your token"}

**Comment by jfayad** [comment]
This is what I'm doing, but as I mentioned it, the token expires after a while and it needs a new login.

This has two major disadvantages:
- I have to store the credentials of a user... if these get hacked all the user's access are compromised
- I had to setup a "retry" mechanism in place for every API call so that a new token will be generated if the one being used is expired... this adds delay to requests

I really think having the ability to generate an API key per user is a more standard approach

**Comment by openpim** [comment]
Hi @jfayad,

Can you please clarify what do you mean  generate an API key per user? Provide login/password with each request? Or opposite use one token for one user forever?

**Comment by jfayad** [comment]
Basically I was expecting to have an API key that a user can generate (and define permissions associated with each key).

Then in the API we could use that Key which wouldn't expire until its revoked by the user.

Moreover, the API key can't be used to generate new keys or to edit the user password

**Comment by SebastiaanDenHertog** [comment]
I get your point, but then you are storing a api key instead of the user details. So if you get hacked you have the same problem. But in theorie what you describes is already a thing you can create a role(the tings a user can and cannot do) and the user with that role. The you login and get your api key. So yes i get that it is more work to implement the login stuff. But safety wise i like the way openPim has implemented it right now.

**Comment by openpim** [comment]
Hi @jfayad 

Current approach to receive the token is standard way that is widely used and I do not see any big reason to change it.
What you can do is to define the time when you token is active. You can set the options to the user  "expiresIn" define the time when token is valid (1d - day, 1h - hour, etc) and also option "passwordChange" = false will prohibit the ability to change the password for this user

![Image](https://github.com/user-attachments/assets/144dc858-83da-4a48-ac6f-c95762dda04a)

**Comment by jfayad** [comment]
these options should definitely be documented somewhere :-)

Any way to have a non expiring token ?
Also, can the option be edited back for the passwordChange ?

**Comment by openpim** [comment]
yes, documentation is a big problem :( We do not have enough resources for this now :(

fully  non expiring token is not available but you can set very bug expiration time.

Option for passwordChange can be changed back.

**Comment by jfayad** [comment]
> yes, documentation is a big problem :( We do not have enough resources for this now :(

We've been using OpenPIM and are very happy with it, how can we help ?

> Option for passwordChange can be changed back.

Is this something the user himself can change ? or is it just a possibility for the admin ?

**Comment by openpim** [comment]
It is important for us that the system is using by others, so you are already helping :) 
And documentation we should write ourselves is in any case.

Regarding the user options update. It can be updated by admin or any other user who has permissions to edit users
